### PR TITLE
AArch64 Debug Info fixes

### PR DESCRIPTION
**Pull request by adinn**
This patch fixes several problems with debug info on AArch64. Associated issue is https://github.com/oracle/graal/issues/2378.

Fixed functionality includes

.debug_frame section CIE records:
  establish correct stack call frame address at entry 

.debug_line section:
  restate initial line position at end of method prologue  -- workaround for bad gdb behaviour

.debug_frame section FDE records:
  notify gdb of save/restore for frame and link register 

In terms of user experience what this achieves is that gdb now
    - places method breaks at the first rather than the second line of a method
    - does not crash when you try to display a stack backtrace
    - correctly identifies the return address and stack top for the previous frame

### CODE REVIEW

File: substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfDebugInfo.java [hunk 1]
```diff
@@ -201,4 +203,8 @@ public DwarfARangesSectionImpl getARangesSectionImpl() {
     public DwarfLineSectionImpl getLineSectionImpl() {
         return dwarfLineSection;
     }
+
+    public ELFMachine getELFMachine() {
```
Comment (olpaw) [inline substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfDebugInfo.java:207]: You don't need an accessor if you make `elfMachine` a public final field.

Comment (adinn) [reply substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfDebugInfo.java]: Sure, no point making the API surface larger than it needs to be. Fixed

**Comment by adinn** [comment]
@olpaw  Hmm, Travis seems to have failed for some spurious reason. Can this be pulled despite that failure?

File: substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfLineSectionImpl.java [hunk 2]
```diff
@@ -478,6 +479,32 @@ private int writeLineNumberTable(DebugContext context, ClassEntry classEntry, by
...
+             * On AArch64 gdb expects to see a line record at the start of the method and a second
+             * one at the end of the prologue marking the point where the method code begins for
+             * real. If we don't provide it then gdb will skip to the second line record when we
+             * place a breakpoint on the method.
+             *
+             * We can identify the end of the prologue for normal methods by noting where the stack
+             * frame height is first adjusted. This should normally be no more a few instructions in
+             * total.
+             */
+            if (isAArch64() && !primaryEntry.getFrameSizeInfos().isEmpty()) {
```
Comment (olpaw) [inline substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfLineSectionImpl.java:492]: If doing the same also for AMD64 does not have any negative side effects then maybe it's not worth to make this piece architecture specific.

**Comment by olpaw** [comment]
> @olpaw Hmm, Travis seems to have failed for some spurious reason. Can this be pulled despite that failure?

Sure. Not sure though if it will merge before Monday.

**Pull request closed without merging**